README.md
# b - Base conversion

[![Base conversion](https://raw.githubusercontent.com/roddamnkit/b/master/img/.b.png)](https://github.com/roddamnkit/b)

[![Version](https://img.shields.io/pub/v/b)](https://pub.dev/packages/b)
[![Build Status](https://github.com/roddamnkit/b/actions/workflows/.github/workflows/dart.yml/badge.svg?branch=master)](https://github.com/roddamnkit/b/actions)
[![Coverage Status](https://coveralls.io/repos/github/roddamnkit/b/badge.svg)](https://coveralls.io/github/roddamnkit/b)
[![License](https://img.shields.io/badge/license-MIT-green)](https://github.com/roddamnkit/b/blob/master/LICENSE)

A Dart library for converting between different bases, e.g., decimal ↔ binary, octal ↔ hexadecimal

## Usage

See `example/main.dart`

#### Hexadecimal ↔ Base58
```dart
final BaseConversion hexadecimalToBase58Converter = BaseConversion(
  from: '0123456789abcdef',
  to: base58,
);

final String base58Value = hexadecimalToBase58Converter('415a59758fb933b6049b050a556dd4d916b7b483f6966615');
// base58Value == '6xZA4Qt9vH7rePWeT5WLaVUZNjB6u6rGc'

final String hexadecimal = hexadecimalToBase58Converter.inverse()('GjWGF6jERR9ymrC1bHcGmsJYkLMDoaySr');
// hexadecimal == 'ac93c8d619c76f823f184110759b278f246cc7cc3cadcac3'
```

#### Decimal (emoji) ↔ Hexadecimal
```dart
final BaseConversion decimalEmojiToHexadecimalConverter = BaseConversion(
  from: '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
  to: base16,
  zeroPadding: true,
);

final String hexadecimal = decimalEmojiToHexadecimalConverter('5️⃣1️⃣9️⃣6️⃣6️⃣');
// hexadecimal == '0CAFE'

final String decimalEmoji = decimalEmojiToHexadecimalConverter.inverse()('DEADC0DE');
// decimalEmoji == '3️⃣7️⃣3️⃣5️⃣9️⃣2️⃣9️⃣0️⃣5️⃣4️⃣'
```

### Exported alphabets
- `base2` - `01`
- `base3` - `012`
- `base4` - `0123`
- `base5` - `01234`
- `base6` - `012345`
- `base8` - `01234567`
- `base10` - `0123456789`
- `base12` - `0123456789AB`
- `base16` - `0123456789ABCDEF`
- `base32` - `ABCDEFGHIJKLMNOPQRSTUVWXYZ234567`
- `base32hex` - `0123456789ABCDEFGHIJKLMNOPQRSTUV`
- `base36` - `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ`
- `base58` - `123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz`
- `base62` - `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz`
- `base64` - `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/`
- `base64url` - `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_`

## Syntax

### *`BaseConversion({required String from, required String to, bool zeroPadding = false})`*

*`from`* - String of numeral symbols representing the digits of `from` numeral system.

*`to`* - String of numeral symbols representing the digits of `to` numeral system.

See [documentation](https://pub.dev/documentation/b) for more

## License

[MIT](https://github.com/roddamnkit/b/blob/master/LICENSE)


CHANGELOG.md
## 0.2.0

- Added zero padding option

## 0.1.2

- Added `base62` alphabet

## 0.1.1+1

- Updated `README.md`

## 0.1.1

- Exported standard alphabets

## 0.1.0

- Initial development release

## 0.1.0-rc.0

- Release candidate for `0.1.0`

## 0.1.0-nameclaim

- Initial development version (name claim), created by Bhargav


test/alphabet_test.dart
import 'package:b/b.dart';
import 'package:test/test.dart';

void main() {
  validationTests();
  radixTests();
  isValidTests();
}

void validationTests() {
  test('alphabet has duplicate characters', () {
    expect(
      () => Alphabet('0123453789'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'alphabet',
          'argument is alphabet',
        ),
      )),
    );
  });

  test('radixPoint has no characters', () {
    expect(
      () => Alphabet('0123456789', radixPoint: ''),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'radixPoint',
          'argument is radixPoint',
        ),
      )),
    );
  });

  test('radixPoint has more than one character', () {
    expect(
      () => Alphabet('0123456789', radixPoint: '..'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'radixPoint',
          'argument is radixPoint',
        ),
      )),
    );
  });

  test('radixPoint is part of alphabet', () {
    expect(
      () => Alphabet('0123456789', radixPoint: '5'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'radixPoint',
          'argument is radixPoint',
        ),
      )),
    );
  });

  test('repetendSymbols has no characters', () {
    expect(
      () => Alphabet('0123456789', repetendSymbols: ''),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'repetendSymbols',
          'argument is repetendSymbols',
        ),
      )),
    );
  });

  test('repetendSymbols has more than two characters', () {
    expect(
      () => Alphabet('0123456789', repetendSymbols: '{()}'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'repetendSymbols',
          'argument is repetendSymbols',
        ),
      )),
    );
  });

  test('a repetendSymbol is part of alphabet', () {
    expect(
      () => Alphabet('0123456789', repetendSymbols: '8'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'repetendSymbols',
          'argument is repetendSymbols',
        ),
      )),
    );
  });

  test('a repetendSymbol and radixPoint are same', () {
    expect(
      () => Alphabet('0123456789', radixPoint: '{', repetendSymbols: '{}'),
      throwsA(allOf(
        isArgumentError,
        predicate(
          (ArgumentError e) => e.name == 'repetendSymbols',
          'argument is repetendSymbols',
        ),
      )),
    );
  });
}

void radixTests() {
  test('base-10', () {
    final Alphabet alphabet = Alphabet('0123456789');
    expect(alphabet.radix, equals(10));
  });

  test('base-2', () {
    final Alphabet alphabet = Alphabet('01');
    expect(alphabet.radix, equals(2));
  });

  test('base-16', () {
    final Alphabet alphabet = Alphabet('0123456789abcdef');
    expect(alphabet.radix, equals(16));
  });

  test('base-10 (emoji)', () {
    final Alphabet alphabet = Alphabet(
      '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
      radixPoint: '*️⃣',
      repetendSymbols: '#️⃣',
    );
    expect(alphabet.radix, equals(10));
  });
}

void isValidTests() {
  test('valid base-10 numerals', () {
    final Alphabet alphabet = Alphabet('0123456789');

    final List<String> validNumerals = <String>[
      '123',
      '.123',
      '123.45',
      '123.45(67)',
      '123.(67)',
    ];

    validNumerals.forEach((String n) => expect(
        '$n is ${alphabet.isValid(n) ? 'valid' : 'invalid'}',
        equals('$n is valid')));
  });

  test('invalid base-10 numerals', () {
    final Alphabet alphabet = Alphabet('0123456789');

    final List<String> invalidNumerals = <String>[
      '',
      '123abc',
      '123.',
      '123(45)',
      '123.4.5',
      '123.45(67)8',
      '123.45(67',
      '123.4567(',
      '123.4567)',
      '123.45)67(',
      '123.45(67(',
      '123.45)67)',
      '123.4567()',
      '123.4567(())',
      '123.4567()()',
    ];

    invalidNumerals.forEach((String n) => expect(
        '$n is ${alphabet.isValid(n) ? 'valid' : 'invalid'}',
        equals('$n is invalid')));
  });

  test('valid base-10 (emoji) numerals', () {
    final Alphabet alphabet = Alphabet(
      '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
      radixPoint: '*️⃣',
      repetendSymbols: '#️⃣',
    );

    final List<String> validNumerals = <String>[
      '2️⃣6️⃣4️⃣8️⃣',
      '*️⃣2️⃣6️⃣4️⃣8️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣#️⃣0️⃣5️⃣#️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣#️⃣0️⃣5️⃣#️⃣',
    ];

    validNumerals.forEach((String n) => expect(
        '$n is ${alphabet.isValid(n) ? 'valid' : 'invalid'}',
        equals('$n is valid')));
  });

  test('invalid base-10 (emoji) numerals', () {
    final Alphabet alphabet = Alphabet(
      '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
      radixPoint: '*️⃣',
      repetendSymbols: '#️⃣',
    );

    final List<String> invalidNumerals = <String>[
      '',
      '2️⃣6️⃣4️⃣8️⃣🔢',
      '2️⃣6️⃣4️⃣8️⃣*️⃣',
      '2️⃣6️⃣4️⃣8️⃣#️⃣0️⃣5️⃣#️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣*️⃣1️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣#️⃣0️⃣5️⃣#️⃣1️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣#️⃣0️⃣5️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣0️⃣5️⃣#️⃣',
      '2️⃣6️⃣4️⃣8️⃣*️⃣9️⃣1️⃣0️⃣5️⃣#️⃣#️⃣',
    ];

    invalidNumerals.forEach((String n) => expect(
        '$n is ${alphabet.isValid(n) ? 'valid' : 'invalid'}',
        equals('$n is invalid')));
  });
}


test/base_conversion_test.dart
import 'package:test/test.dart';

import 'decimal_to_binary_testcase.dart';
import 'decimal_to_decimal_emoji_testcase.dart';

void main() {
  testConverter(decimalToBinary);
  testConverter(decimalToDecimalEmoji);
}

void testConverter(Map<String, dynamic> converter) {
  group('function', () {
    converter['testCases'].forEach((Map<String, String> testCase) {
      test('${testCase['from']} -> ${testCase['to']}', () {
        expect(converter['instance'](testCase['from']), equals(testCase['to']));
      });
    });
  });

  group('inverse function', () {
    converter['testCases'].forEach((Map<String, String> testCase) {
      test('${testCase['to']} -> ${testCase['from']}', () {
        expect(converter['instance'].inverse()(testCase['to']),
            equals(testCase['from']));
      });
    });
  });
}


test/decimal_to_binary_testcase.dart
import 'package:b/b.dart';

final Map<String, dynamic> decimalToBinary = <String, dynamic>{
  'instance': BaseConversion(
    from: decimal,
    to: binary,
  ),
  'testCases': <Map<String, String>>[
    <String, String>{'from': '0', 'to': '0'},
    <String, String>{'from': '1', 'to': '1'},
    <String, String>{'from': '2', 'to': '10'},
    <String, String>{'from': '3', 'to': '11'},
    <String, String>{'from': '4', 'to': '100'},
    <String, String>{'from': '5', 'to': '101'},
    <String, String>{'from': '6', 'to': '110'},
    <String, String>{'from': '7', 'to': '111'},
    <String, String>{'from': '8', 'to': '1000'},
    <String, String>{'from': '9', 'to': '1001'},
    <String, String>{'from': '10', 'to': '1010'},
  ],
};


example/main.dart
import 'package:b/b.dart';

void main() {
  decimalAndBinary();
  decimalAndDecimalEmoji();
  decimalEmojiAndHexadecimal();
}

void decimalAndBinary() {
  // Decimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
  // Binary alphabet - 0, 1
  final BaseConversion decimalToBinaryConverter = BaseConversion(
    from: base10,
    to: base2,
  );

  final String binary = decimalToBinaryConverter('3888');
  print(
      'decimal number 3888 is represented as $binary in binary'); // 111100110000

  final String decimal = decimalToBinaryConverter.inverse()('111110011111');
  print(
      'binary number 111110011111 is represented as $decimal in decimal'); // 3999
}

void decimalAndDecimalEmoji() {
  // Decimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
  // Decimal emoji alphabet - 0️⃣, 1️⃣, 2️⃣, 3️⃣, 4️⃣, 5️⃣, 6️⃣, 7️⃣, 8️⃣, 9️⃣
  final BaseConversion decimalToDecimalEmojiConverter = BaseConversion(
    from: base10,
    to: '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
  );

  final String decimalEmoji = decimalToDecimalEmojiConverter('38888');
  print(
      'decimal number 38888 is represented as $decimalEmoji in decimal emoji'); // 3️⃣8️⃣8️⃣8️⃣8️⃣

  final String decimal =
      decimalToDecimalEmojiConverter.inverse()('3️⃣9️⃣9️⃣9️⃣9️⃣');
  print(
      'decimal emoji number 3️⃣9️⃣9️⃣9️⃣9️⃣ is represented as $decimal in decimal'); // 39999
}

void decimalEmojiAndHexadecimal() {
  // Decimal emoji alphabet - 0️⃣, 1️⃣, 2️⃣, 3️⃣, 4️⃣, 5️⃣, 6️⃣, 7️⃣, 8️⃣, 9️⃣
  // Hexadecimal alphabet - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F
  final BaseConversion decimalEmojiToHexadecimalConverter = BaseConversion(
    from: '0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣',
    to: base16,
    zeroPadding: true,
  );

  final String hexadecimal =
      decimalEmojiToHexadecimalConverter('5️⃣1️⃣9️⃣6️⃣6️⃣');
  print(
      'decimal emoji number 5️⃣1️⃣9️⃣6️⃣6️⃣ is represented as $hexadecimal in hexadecimal'); // 0CAFE

  final String decimalEmoji =
      decimalEmojiToHexadecimalConverter.inverse()('DEADC0DE');
  print(
      'hexadecimal number DEADC0DE is represented as $decimalEmoji in decimal emoji'); // 3️⃣7️⃣3️⃣5️⃣9️⃣2️⃣9️⃣0️⃣5️⃣4️⃣
}


lib/src/base_conversion.dart
part of b;

/// Provides ability to convert numbers between different
/// positional numeral systems.
class BaseConversion extends InvertibleFunction<String, String> {
  /// Constructs a `BaseConversion` object with the given [from] and [to] alphabets.
  BaseConversion({
    required String from,
    required String to,
    bool zeroPadding = false,
  }) : this.alphabet(
          from: Alphabet(from),
          to: Alphabet(to),
          zeroPadding: zeroPadding,
        );

  /// Constructs a `BaseConversion` object with the given [from] and [to] alphabets.
  BaseConversion.alphabet({
    required Alphabet from,
    required Alphabet to,
    bool zeroPadding = false,
  })  : _fromAlphabet = from,
        _toAlphabet = to,
        _lengthFactor = __lengthFactor(from: from, to: to),
        _intLengthFactor = (() {
          final double lengthFactor = __lengthFactor(from: from, to: to);
          int intLengthFactor = lengthFactor.truncate();
          if (lengthFactor == intLengthFactor) {
            return intLengthFactor;
          } else {
            final double invLengthFactor = (1 / lengthFactor);
            intLengthFactor = -invLengthFactor.truncate();
            if (invLengthFactor == -intLengthFactor) {
              return intLengthFactor;
            }
          }
          return 0;
        })(),
        _zeroPadding = zeroPadding;

  final Alphabet _fromAlphabet;
  final Alphabet _toAlphabet;

  final double _lengthFactor;
  final int _intLengthFactor;

  final bool _zeroPadding;

  @override
  List<bool Function(String)> get domain =>
      <bool Function(String)>[_fromAlphabet.isValid];

  @override
  String valueAt(String x) {
    final List<String> parts = x.split(_fromAlphabet._radixCharacter);
    String r = _convertIntegralPart(parts[0]);
    if (parts.length > 1) {
      r += _toAlphabet._radixCharacter + _convertFractionalPart(parts[1]);
    }
    return r;
  }

  String _convertIntegralPart(String ip) {
    final int fromBase = _fromAlphabet.radix;
    final int toBase = _toAlphabet.radix;

    (String, int) changeBase(Iterable<int> _values) {
      // Horner's method
      final Iterable<int> values = _values.skipWhile((int v) => v == 0);

      if (values.isEmpty) {
        return ('', 0);
      }

      int remainder = 0;
      final List<int> quotients = <int>[];
      for (final int value in values) {
        remainder = (remainder * fromBase) + value;

        quotients.add(remainder ~/ toBase);
        remainder = remainder % toBase;
      }

      final (String r, int rLen) = changeBase(quotients);

      return (r + _toAlphabet._characters[remainder], rLen + 1);
    }

    final int ipLen = ip.characters.length;

    var (String r, int rLen) =
        changeBase(ip.characters.map(_fromAlphabet._characters.indexOf));

    if (_zeroPadding) {
      final int currentLength = rLen;
      final int wantedLength = (ipLen * _lengthFactor).ceil();

      if (currentLength < wantedLength) {
        final StringBuffer sb = StringBuffer();
        sb.write(_toAlphabet._zeroCharacter * (wantedLength - currentLength));
        sb.write(r);

        r = sb.toString();
      }
    } else if (r.isEmpty) {
      r = _toAlphabet._zeroCharacter;
    }

    return r;
  }

  // TODO
  String _convertFractionalPart(String fp) => _toAlphabet._zeroCharacter;

  @override
  BaseConversion inverse() => BaseConversion.alphabet(
        from: _toAlphabet,
        to: _fromAlphabet,
        zeroPadding: _zeroPadding,
      );
}

double __lengthFactor({required Alphabet from, required Alphabet to}) =>
    log(from.radix) / log(to.radix);


lib/src/alphabet.dart
part of b;

///
class Alphabet {
  ///
  Alphabet(
    String alphabet, {
    String radixPoint = '.',
    String repetendSymbols = '()',
  })  : _characters = alphabet.characters.toList(),
        _zeroCharacter = alphabet.characters.first,
        _radixCharacter = radixPoint.characters.length == 1
            ? radixPoint.characters.first
            : throw ArgumentError.value(
                radixPoint,
                'radixPoint',
                'Must have one and only one character for radix point',
              ),
        _repetendStartCharacter = repetendSymbols.characters.length == 1 ||
                repetendSymbols.characters.length == 2
            ? repetendSymbols.characters.first
            : throw ArgumentError.value(
                repetendSymbols,
                'repetendSymbols',
                'Must have one and only one character for a repetend symbol',
              ),
        _repetendEndCharacter = repetendSymbols.characters.length == 1 ||
                repetendSymbols.characters.length == 2
            ? repetendSymbols.characters.last
            : throw ArgumentError.value(
                repetendSymbols,
                'repetendSymbols',
                'Must have one and only one character for a repetend symbol',
              ) {
    if (_characters.length != _characters.toSet().length) {
      throw ArgumentError.value(
        alphabet,
        'alphabet',
        'Must not have duplicate characters in alphabet',
      );
    }
    if (_characters.contains(_radixCharacter)) {
      throw ArgumentError.value(
        radixPoint,
        'radixPoint',
        'Radix point must not be part of alphabet',
      );
    }
    if (_characters.contains(_repetendStartCharacter) ||
        _radixCharacter == _repetendStartCharacter ||
        _characters.contains(_repetendEndCharacter) ||
        _radixCharacter == _repetendEndCharacter) {
      throw ArgumentError.value(
        repetendSymbols,
        'repetendSymbols',
        'A repetend symbol must not be part of alphabet and be different from the radix point',
      );
    }
  }

  final List<String> _characters;
  final String _zeroCharacter;

  final String _radixCharacter;
  final String _repetendStartCharacter;
  final String _repetendEndCharacter;

  /// Radix of this alphabet
  int get radix => _characters.length;

  /// Checks if [s] is a valid numeral representation
  bool isValid(String s) {
    String? previousCharacter;
    bool foundRadixCharacter = false;
    bool foundRepetendStartCharacter = false;
    bool foundRepetendEndCharacter = false;
    for (final String character in s.characters) {
      if (!_characters.contains(character) &&
          _radixCharacter != character &&
          _repetendStartCharacter != character &&
          _repetendEndCharacter != character) {
        return false;
      }
      if (_radixCharacter == character) {
        if (!foundRadixCharacter) {
          foundRadixCharacter = true;
        } else {
          return false;
        }
      } else if (_repetendStartCharacter == character ||
          _repetendEndCharacter == character) {
        if (_repetendStartCharacter == character &&
            foundRadixCharacter &&
            !foundRepetendStartCharacter) {
          foundRepetendStartCharacter = true;
        } else if (_repetendEndCharacter == character &&
            foundRepetendStartCharacter &&
            !foundRepetendEndCharacter &&
            _characters.contains(previousCharacter)) {
          foundRepetendEndCharacter = true;
        } else {
          return false;
        }
      }
      previousCharacter = character;
    }
    return previousCharacter != null &&
        previousCharacter != _radixCharacter &&
        (!foundRepetendStartCharacter ||
            (foundRepetendEndCharacter &&
                previousCharacter == _repetendEndCharacter));
  }
}

/// Binary numeral system
const String binary = '01';

/// Same as [binary] numeral system
const String base2 = binary;

/// Ternary numeral system
const String ternary = '012';

/// Same as [ternary] numeral system
const String base3 = ternary;

/// Quaternary numeral system
const String quaternary = '0123';

/// Same as [quaternary] numeral system
const String base4 = quaternary;

/// Quinary numeral system
const String quinary = '01234';

/// Same as [quinary] numeral system
const String base5 = quinary;

/// Senary numeral system
const String senary = '012345';

/// Same as [senary] numeral system
const String base6 = senary;

/// Octal numeral system
const String octal = '01234567';

/// Same as [octal] numeral system
const String base8 = octal;

/// Decimal numeral system
const String decimal = '0123456789';

/// Same as [decimal] numeral system
const String base10 = decimal;

/// Duodecimal numeral system
const String duodecimal = '0123456789AB';

/// Same as [duodecimal] numeral system
const String base12 = duodecimal;

/// Hexadecimal numeral system
const String hexadecimal = '0123456789ABCDEF';

/// Same as [hexadecimal] numeral system
const String base16 = hexadecimal;

/// Base32 numeral system
const String base32 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

/// Base32 (Extended Hex) numeral system
const String base32hex = '0123456789ABCDEFGHIJKLMNOPQRSTUV';

/// Base36 numeral system
const String base36 = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';

/// Base58 numeral system
const String base58 =
    '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/// Base62 numeral system
const String base62 =
    '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

/// Base64 numeral system
const String base64 =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

/// Base64 (URL and Filename safe) numeral system
const String base64url =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
